pytools/graph.py

```diff
@@ -369,4 +370,54 @@ def compute_induced_subgraph(graph: Mapping[T, Set[T]],
 
 # }}}
 
+
+# {{{
+
+def visualize_graph(graph: Mapping[T, Iterable[T]],
+                    output_to: Optional[str] = None) -> str:
+    """
+    Visualize the graph *graph*. Needs the ``dot`` program installed.
+
+    :arg graph: A :class:`collections.abc.Mapping` representing a directed
+        graph. The dictionary contains one key representing each node in the
+        graph, and this key maps to a :class:`collections.abc.Set` of nodes
+        that are connected to the node by outgoing edges.
+
+    :arg output_to: See :func:`pytools.dot.show_dot` for details.
+
+    :returns: A string in the `dot <http://graphviz.org/>`__ language.
+    """
+    from pytools import UniqueNameGenerator
+    id_gen = UniqueNameGenerator(forced_prefix="mynode")
+
+    from pytools.dot import dot_escape, show_dot
+
+    node_to_id = {}
+    edges = []
+
+    for node, targets in graph.items():
+        if node not in node_to_id:
+            node_to_id[node] = id_gen()
+        for t in targets:
+            if t not in node_to_id:
+                node_to_id[t] = id_gen()
+            edges.append((node, t))
+
+    content = "\n".join(
+        [f'{node_to_id[node]} [label="{dot_escape(repr(node))}"];'
+         for node in node_to_id.keys()])
+
+    content += "\n"
+
+    content += "\n".join(
+        [f"{node_to_id[node]} -> {node_to_id[t]};" for (node, t) in edges])
+
+    res = f"digraph mygraph {{\n{ content }\n}}\n"
+
+    show_dot(res, output_to)
```

Comment:
Calling an external executable here doesn't seem necessary, since
```
show_dot(visualize_graph(...), output_to="svg")
```
is pretty short. Any specific reason to bundle these together?

Reply:
My main motivation was based on my impression that few users actually want to "get" the dot code, but rather want to visualize it. But you're right, it isn't too much extra code, and makes the test simpler, so I'll change it to `get_graph_dot_code` and remove the call to `show_dot`.

Reply:
That's fair enough. I'm not a user of most of this stuff, so take my comment with a sizeable grain of salt!

(The use case I had in mind was just writing the dot code to a file for later)

Reply:
No worries, I think the API is much cleaner now! Thanks for the suggestion!